src/Generator/Passes/DelegatesPass.cs

```diff
@@ -35,33 +43,135 @@ public DelegatesPass()
+                }
+
+                if (Options.IsCSharpGenerator)
+                    return AddDelegateToDictionary(@namespace, decl);
+
+                var delegateName = GenerateDelegateSignature(functionType.Parameters, functionType.ReturnType);
+
+                typedef = new TypedefDecl
+                {
+                    Access = AccessSpecifier.Public,
+                    Name = delegateName,
+                    Namespace = @namespace,
+                    QualifiedType = type,
+                    IsSynthetized = true
+                };
+                typedefs.Add(new Typedef
+                {
+                    Context = @namespace,
+                    Declaration = typedef
+                });
+            }
+
+            var typedefType = new TypedefType
+            {
+                Declaration = typedef
+            };
+            return new QualifiedType(typedefType);
+        }
+
         public override bool VisitASTContext(ASTContext context)
         {
-            foreach (var module in Options.Modules)
-                libsDelegates[module] = new Dictionary<string, DelegateDefinition>();
+            bool result = base.VisitASTContext(context);
 
-            var unit = context.TranslationUnits.GetGenerated().LastOrDefault();
+            foreach (var typedef in allTypedefs)
+            {
+                foreach (var foo in typedef.Value)
+                {
+                    foo.Context.Declarations.Add(foo.Declaration);
+                }
+            }
+            allTypedefs.Clear();
 
-            if (unit == null)
-                return false;
+            if (Options.IsCSharpGenerator)
+            {
+                foreach (var module in Options.Modules)
+                    libsDelegates[module] = new Dictionary<string, DelegateDefinition>();
+
+                var unit = context.TranslationUnits.GetGenerated().LastOrDefault();
 
-            var result = base.VisitASTContext(context);
+                if (unit == null)
+                    return false;
 
-            foreach (var module in Options.Modules.Where(m => namespacesDelegates.ContainsKey(m)))
-                namespacesDelegates[module].Namespace.Declarations.Add(namespacesDelegates[module]);
+                foreach (var module in Options.Modules.Where(m => namespacesDelegates.ContainsKey(m)))
+                    namespacesDelegates[module].Namespace.Declarations.Add(namespacesDelegates[module]);
+            }
 
             return result;
         }
 
-        private void AddDelegatesToDictionary(Declaration decl, Module module)
+        private QualifiedType AddDelegateToDictionary(DeclarationContext @namespace, Declaration decl)
         {
             CallingConvention callingConvention;
             bool isDependent = false;
             QualifiedType returnType;
             List<Parameter> @params;
-            if (decl is Method)
+            bool isVirtualMethod = false;
```

Comment:
Not sure what isVirtualMethod is for. If it's a method, we have a delegate for a virtual function, if not, it's for a parameter so no method at all.